### DISCUSSION TITLE
Trouble Sending RPC to Create Table

### DISCUSSION THREAD

**Discussion by syamsulalam** [supabase/supabase#20711] [Questions] [answered]
so basically i got the following payload to send to supabase (i try to create new table from my javascript hosted in github page)

```{table_name: 'testing', columns: Array(4)} columns: (4) ['hewan', 'gila', 'bekicot', 'imut']
table_name:"testing"
[[Prototype]]:Object
```

i try to send it through RPC to my Postgres function but it kept on failing:

`POST https://mwikqvfpuxttqjucmhoj.supabase.co/rest/v1/rpc/create_table 400 (Bad Request)`

My postgres function / query are as follows:

```sql
CREATE OR REPLACE FUNCTION public.create_table (
  table_name character varying,
  columns character varying[]
) RETURNS void LANGUAGE 'plpgsql' SECURITY DEFINER AS $BODY$
DECLARE
    column_declaration text;
BEGIN
    SET search_path = public;
    -- Create the column declaration string
    column_declaration := array_to_string(columns, ', ');

    -- Create the table
    EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')';
END;
$BODY$;
```

i kept on getting 42601 error which i know it is syntax error but i can't see where is my error:

```{data: null, error: {…}}
  data: null
    error: 
      code: "42601"
      details: null
      hint: null
      message: "syntax error at end of input"
      [[Prototype]]: Object
[[Prototype]]: Object
```

what to do now? thanks. my codes are here: https://cekkode.github.io/sheets/supabase.html

**Comment by GaryAustin1** [comment]
This should be a pure Postgres SQL issue based on your error.  Possibly related to your column string, but not clear.
You should research on Google/Stackoverflow or even ask an AI as they do well at pure SQL.   
There is nothing specific to Supabase here, so you are relying on an SQL guru to find this or another user to do what you could do and research/debug.

You do have a plpgsql function so you can use `raise log 'val = 5',column_declaration;` and see the result in the Postgres log.

**Reply by syamsulalam** [reply, marked as answer]
thanks for pointing to the right direction. i modify the function as you suggested like this:

```sql
CREATE OR REPLACE FUNCTION public.create_table (
  table_name character varying,
  columns character varying[]
) RETURNS void LANGUAGE 'plpgsql' SECURITY DEFINER AS $BODY$
DECLARE
    column_declaration text;
BEGIN
    SET search_path = public;
    -- Create the column declaration string
    column_declaration := array_to_string(columns, ', ');

    -- Log information for debugging
    RAISE LOG 'Creating table: %', table_name;
    RAISE LOG 'Column declaration: %', column_declaration;

    -- Create the table using EXECUTE
    EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')';

END;
$BODY$;
```
which still has 42601 error, but then after copy & paste the postgres logs to chatgpt, lo and behold, chatgpt able to found out the problem and fix the function to this:

```sql
-- Supabase AI is experimental and may produce incorrect answers
-- Always verify the output before executing

CREATE OR REPLACE FUNCTION public.create_table (
  table_name character varying,
  columns character varying[]
) RETURNS void LANGUAGE 'plpgsql' SECURITY DEFINER AS $BODY$
DECLARE
    column_declaration text;
BEGIN
    SET search_path = public;
    -- Create the column declaration string
    column_declaration := array_to_string(columns, ' text, ') || ' text';

    -- Log information for debugging
    RAISE LOG 'Creating table: %', table_name;
    RAISE LOG 'Column declaration: %', column_declaration;

    -- Create the table using EXECUTE
    EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')';

END;
$BODY$;
```

and now it works perfectly. thank you.

**Comment by steve-chavez** [comment]
Never use dynamic SQL like that or you'll be vulnerable to SQL injection. Use the `EXECUTE .. USING` form, see https://www.postgresql.org/docs/current/plpgsql-statements.html#PLPGSQL-STATEMENTS-EXECUTING-DYN.

**Reply by syamsulalam** [reply]
i try to implement your suggestion and modify my function like this:

```sql
CREATE OR REPLACE FUNCTION public.create_table (
  table_name character varying,
  columns character varying[]
) RETURNS void LANGUAGE 'plpgsql' SECURITY DEFINER AS $BODY$
DECLARE
    column_declaration text;
BEGIN
    SET search_path = public;
    -- Create the column declaration string
    column_declaration := array_to_string(columns, ', ');

    -- Log information for debugging
    RAISE LOG 'Creating table: %', table_name;
    RAISE LOG 'Column declaration: %', column_declaration;

    -- Create the table using EXECUTE with USING
    EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')' USING;

END;
$BODY$;
```
but then i got this error:

```
ERROR:  42601: missing expression at or near ";"
LINE 17:     EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')' USING;
                                                                                             ^
```

then i ask chatgpt about it:

> I apologize for the confusion. It seems there was a misunderstanding in my previous responses. The USING clause is not applicable in the context of the CREATE TABLE statement because CREATE TABLE doesn't involve parameterized values in the same way as a typical query with parameters.
> 
> To clarify, the USING clause is commonly used with the EXECUTE statement when you have parameters in your SQL query that need to be safely inserted to prevent SQL injection. However, in the case of CREATE TABLE, the statement involves the table name and column definitions, which are not parameterized in the same way as values in a typical SQL query.
> 
> Here is the correct version of your function without the USING clause:

```sql
CREATE OR REPLACE FUNCTION public.create_table (
  table_name character varying,
  columns character varying[]
) RETURNS void LANGUAGE 'plpgsql' SECURITY DEFINER AS $BODY$
DECLARE
    column_declaration text;
BEGIN
    SET search_path = public;
    -- Create the column declaration string
    column_declaration := array_to_string(columns, ', ');

    -- Log information for debugging
    RAISE LOG 'Creating table: %', table_name;
    RAISE LOG 'Column declaration: %', column_declaration;

    -- Create the table using EXECUTE
    EXECUTE 'CREATE TABLE ' || table_name || ' (' || column_declaration || ')';

END;
$BODY$;
```
> I appreciate your patience, and I apologize for any confusion caused by the oversight in my previous responses.

**Reply by GaryAustin1** [reply]
Note for other's considering this:

I've discussed this with the OP in Discord before posting earlier.  But for anyone else coming along,  generating tables on the fly like this is consider anti-pattern and bad practice, except in special cases.  The OP may have such a case.  But often users want to do this as a table per user and this is a bad idea.

Also, in addition to Steve's comment I had mentioned to the OP this function is not secure and anyone could call it.  Some sort of grant removal or testing for user roles is normally needed for a security definer function.  Also in generating a table with simple SQL like this, RLS was not addressed which also means the table is vulnerable thru the API.
